src/storage/src/render/persist_sink.rs

```diff
@@ -932,439 +935,574 @@ where
+            &mut in_flight_descriptions,
+        )
+        .await;
+
+        if let Err(e) = result {
+            return Err(e);
+        }
+    }
+}
 
-            while let Some(done_batch_metadata) = done_batches.pop() {
-                in_flight_descriptions.remove(&done_batch_metadata);
+async fn update_persist_with_batches(
+    collection_id: GlobalId,
+    worker_id: usize,
+    shard_id: ShardId,
+    metrics: &SourcePersistSinkMetrics,
+    busy_signal: &Arc<Semaphore>,
+    read_only_rx: &mut watch::Receiver<bool>,
+    bail_on_concurrent_modification: bool,
+    write: &mut mz_persist_client::write::WriteHandle<
+        SourceData,
+        (),
+        mz_repr::Timestamp,
+        StorageDiff,
+    >,
+    upper_cap_set: &mut CapabilitySet<mz_repr::Timestamp>,
+    current_upper: &Rc<std::cell::RefCell<Antichain<mz_repr::Timestamp>>>,
+    source_statistics: &StorageStatistics<
+        SourceStatisticsRecord,
+        SourceStatisticsMetrics,
+        SourceStatisticsMetadata,
+    >,
+    done_batches: &mut Vec<(Antichain<mz_repr::Timestamp>, Antichain<mz_repr::Timestamp>)>,
+    in_flight_batches: &mut HashMap<
+        (Antichain<mz_repr::Timestamp>, Antichain<mz_repr::Timestamp>),
+        BatchSet,
+    >,
+    in_flight_descriptions: &mut std::collections::HashSet<(
+        Antichain<mz_repr::Timestamp>,
+        Antichain<mz_repr::Timestamp>,
+    )>,
+) -> Result<(), anyhow::Error> {
+    let (mut batch_lower, batch_upper) = match (done_batches.first(), done_batches.last()) {
+        (Some(first), Some(last)) => (first.0.clone(), last.1.clone()),
+        (None, None) => {
+            upper_cap_set.downgrade(current_upper.borrow().iter());
+            //
+            return Ok(());
+        }
+        // SAFETY(ptravers): if first exists so must last.
+        (Some(_), None) | (None, Some(_)) => {
+            unreachable!("the list of batches must have a first and last if it is non empty.")
+        }
+    };
+
+    let mut batches: Vec<FinishedBatch> = vec![];
+    let mut batch_metrics: BatchMetrics = BatchMetrics::default();
+    let num_batches = u64::cast_from(done_batches.len());
+
+    for batch_metadata in done_batches.drain(..) {
+        in_flight_descriptions.remove(&batch_metadata);
+        let mut batch = in_flight_batches
+            .remove(&batch_metadata)
+            .unwrap_or_default();
+        batches.append(&mut batch.finished);
+        batch_metrics.add_assign(&batch.batch_metrics);
+    }
 
-                let batch_set = in_flight_batches
-                    .remove(&done_batch_metadata)
-                    .unwrap_or_default();
+    let mut to_append = batches.iter_mut().map(|b| &mut b.batch).collect::<Vec<_>>();
+
+    // NB(ptravers): We loop here to handle competing writers to the same shard.
+    // We handle failure to write in three situations:
+    //  1. the lower of the batch is greater than the current upper of shard - persist_sink exits with an error.
```

Comment:
Can this happen? I thought the changes in the API call of persist allow that.

Also, we should us the "beyond" terminology when talking about comparing frontiers. Specifically we use "f1 is beyond f2" if `PartialOrder::less_equal(f2, f1)` and "f1 is not beyond f2"  if `!PartialOrder::less_equal(f2, f1)`. More rarely we use the term "f1 is strictly beyond f2" to mean `ParialOrder::less_equal(f2, f1) && f1 != f2`

So my question here is if persist refuses the API call if for any of the to-be-committed batches `batch.lower` is strictly beyond the current shard upper. My understanding is that it will not refuse it but maybe I misunderstood the comment.

Reply:
I believe the assertion that would fail here is that we are asserting the current upper of the shard is the lower of the batch and persist will error as the current upper is less than the batch lower.

see: https://github.com/MaterializeInc/materialize/blob/367c04ad36142f617b1aeb69d830f26fd57fb3c2/src/persist-client/src/write.rs#L644

Reply:
I see. I think one confusion here is using the term "batch lower" to refer to the expected shard upper passed to the function. I got confused because in the function call we are also passing in a list of batches and the lower of *those* doesn't have any constraints, right?